### ISSUE TITLE
Timestamping public PGP keys

### ISSUE THREAD

**Issue by snap9** [opentimestamps/opentimestamps-client#70]
I want to timestamp my public PGP key. OpenTimestamps lets me do that. This will allow me to prove to future acquaintances that I did indeed create my public PGP key at a particular time in the past. This is great.

But here's the thing. I also want to prove to these future acquaintances that I created **no other** public PGP keys around a similar time - or at least that I didn't create more than hundreds (or thousands) of other public PGP keys around a similar time without expending millions (or billions) of dollars in computing power.

Is this possible? I wonder if a public PGP key can be attached with information proving that a certain computing power was performed to create it.

I realize this is not directly related to OpenTimestamps, but I'm hoping you guys have ideas on this.

TLDR, I want to show to future folks that a particular **and no other** PGP was mine in the past. I'm hoping I can do better than using trusted third parties for this.

**Comment by snap9** [comment]
@petertodd Any ideas?

Suppose you use a timestamped PGP key to prove you had that PGP key 10 years ago.

But that's not really useful because you could have created thousands of PGP keys back then and today you just conveniently happen to be picking one of those PGP keys to suit your needs.

Similar argument holds for any signed PGP messages. You could have signed thousands of them in the past.

**Comment by petertodd** [comment]
You are correct in thinking that timestamping alone won't do what you're trying to do. In fact, strictly speaking what you're trying to do is impossible. But you can approximate it, by e.g. making it provably expensive for you to have created multiple keys in a particular way, e.g. via proof-of-sacrifice. The way that works is you link the creation of the key to an irrevocable sacrifice, such as sending BTC to an unspendable OP_RETURN output that committed to that PGP key. You can also do things like publish the fact that you created the key in a public place, such as a Twitter account.

**Comment by snap9** [comment]
@petertodd Thanks for these insights!

One more question about **time accuracy**. From what I understand, we're limited to time accuracy of a few hours.

Is there no technology on this planet that can timestamp with more accuracy? Does the Litecoin blockchain give us more time accuracy given its 3 minutes block time (vs. Bitcoin's 10 minutes)? Shouldn't that be at least three times the accuracy?

**Comment by petertodd** [comment]
You can definitely timestamp with more accuracy using trusted timestamping, which is something I'd like OpenTimestamps to eventually support.

Just using a system with a shorter block interval doesn't help you here: the cause of the inaccuracy isn't the time between blocks, but rather what time miners can set and still get away with it.

The problem with decentralized timestamping is essentially that it's really difficult to actually enforce the accuracy of the time field in your blocks; if anything, trying to enforce really strict time requirements creates opportunities for attacks. So as a decentralized consensus system designer, you actually want the loosest time requirements possible, not stricter. Ethereum as an example tries to enforce relatively strict requirements, but that increases the risk of consensus failure and attacks for the whole system.

Not to mention that you want a system that'll still work even if miners get their clocks wrong! Accidents happen - I once missed the start of a conference because my cellphone clock was off, and as far as I could tell the root cause was that the local cellphone network was giving out the wrong time.

**Comment by snap9** [comment]
@petertodd Interesting. Looks like with decentralized timestamping will always have low accuracy. Can you point to some of your favorite tools/software/organizations/companies for trusted timestamping? (Until OpenTimestamps supports it, that is.)

**Comment by petertodd** [comment]
There just aren't very good options for trusted timestamping unfortunately. Guardtime is probably the best in terms of tech, but they're designed to be a commercial product, not a opensource thing.

**Comment by snap9** [comment]
@petertodd Do you have written down thoughts anywhere about how such trusted timestamping should work? I'm curious to learn more. I mean really, if we can do timestamping with second or even minute accuracy it could be revolutionary. Yet I'm not sure how you're going to achieve this :)

**Comment by petertodd** [comment]
I have! See https://lists.opentimestamps.org/pipermail/ots-dev/2017-May/000001.html

**Comment by snap9** [comment]
@petertodd Very interesting. Any ideas of ETA for this trusted timestamps solution?

I don't know Python yet so I can't help. Do you think you'll stick with Python for OpenTimestamps? (I noticed you used pseudo-Rust in that post, so I wonder if you're planning on porting to Rust at some point.)

**Comment by petertodd** [comment]
Probably will be a few more months; I have a few higher priority things that need to be done first.

There will always be a Python OpenTimestamps client, but I am intending to make the rust OpenTimestamps library be the OTS standard, and move at least the calendar server implementation to Rust.

**Comment by jim-obrien-orig** [comment]
This is only relevant to a few comments in this issue, and I believe to the list posting, which proposes a Merkle tree structure. There is a new DLT [(Radix)](https://www.radixdlt.com/) using [Lamport Timestamps](https://en.wikipedia.org/wiki/Lamport_timestamps) with [Merkle trees](https://en.wikipedia.org/wiki/Merkle_tree) as temporal proof. It does not prove the time of an event, but the order of events in a distributed ledger, and appears to solve the double-spend problem in a scalable fashion, using logical time as the basis for consensus. Here is the relevant portion of the whitepaper: https://papers.radixdlt.com/tempo/latest/#temporal-proof-provisioning. (I am not an investor, employee, or coin holder of Radix).
An encryption algorithm that can only be used once to generate a pub/private keypair is an interesting idea though. It seems impossible, but you never know.
